File: cplex/examples/src/cpp/etsp.cpp
// -------------------------------------------------------------- -*- C++ -*-
// File: etsp.cpp
// Version 12.6.1  
// --------------------------------------------------------------------------
// Licensed Materials - Property of IBM
// 5725-A06 5725-A29 5724-Y48 5724-Y49 5724-Y54 5724-Y55 5655-Y21
// Copyright IBM Corporation 2000, 2014. All Rights Reserved.
//
// US Government Users Restricted Rights - Use, duplication or
// disclosure restricted by GSA ADP Schedule Contract with
// IBM Corp.
// --------------------------------------------------------------------------
//
// etsp.cpp -  Solving an earliness-tardiness scheduling problem
//             using CPLEX linearization capabilities.
//                   
// A command line argument is required to run this example.
//
// Example:
//     etsp ../../../examples/data/etsp.dat
//

#include <ilcplex/ilocplex.h>

ILOSTLBEGIN

typedef IloArray<IloIntArray>    IntMatrix;
typedef IloArray<IloNumArray>    NumMatrix;
typedef IloArray<IloNumVarArray> NumVarMatrix;

const IloInt Horizon = 10000;

int
main(int argc, char** argv)
{
   IloEnv env;
   try {
      IloInt i, j;

      const char* filename;
      if (argc > 1)
         filename = argv[1];
      else
         filename = "../../../examples/data/etsp.dat";
      ifstream f(filename, ios::in);
      if (!f) {
         cerr << "No such file: " << filename << endl;
         throw(1);
      }

      IntMatrix   activityOnAResource(env);
      NumMatrix   duration(env);
      IloNumArray jobDueDate(env);
      IloNumArray jobEarlinessCost(env);
      IloNumArray jobTardinessCost(env);

      f >> activityOnAResource;
      f >> duration;
      f >> jobDueDate;
      f >> jobEarlinessCost;
      f >> jobTardinessCost;

      IloInt nbJob      = jobDueDate.getSize();
      IloInt nbResource = activityOnAResource.getSize();

      IloModel model(env);

      // Create start variables
      NumVarMatrix s(env, nbJob);
      for (j = 0; j < nbJob; j++) {
         s[j] = IloNumVarArray(env, nbResource, 0.0, Horizon);
      }

      // State precedence constraints
      for (j = 0; j < nbJob; j++) {
         for (i = 1; i < nbResource; i++) {
            model.add(s[j][i] >= s[j][i-1] + duration[j][i-1]);
         }
      }

      // State disjunctive constraints for each resource
      for (i = 0; i < nbResource; i++) {
         IloInt end = nbJob - 1;
         for (j = 0; j < end; j++) {
            IloInt a = activityOnAResource[i][j];
            for (IloInt k = j + 1; k < nbJob; k++) {
              IloInt b = activityOnAResource[i][k];
              model.add(s[j][a] >= s[k][b] + duration[k][b] ||
                        s[k][b] >= s[j][a] + duration[j][a]);
            }
         }
      }

      // The cost is the sum of earliness or tardiness costs of each job 
      IloInt last = nbResource - 1;
      IloExpr costSum(env);
      for (j = 0; j < nbJob; j++) {
         costSum += IloPiecewiseLinear(s[j][last] + duration[j][last],
            IloNumArray(env, 1, jobDueDate[j]),
            IloNumArray(env, 2, -jobEarlinessCost[j], jobTardinessCost[j]),
            jobDueDate[j], 0);
      }
      model.add(IloMinimize(env, costSum));
      costSum.end();

      IloCplex cplex(env);

      cplex.extract(model);

      cplex.setParam(IloCplex::Param::Emphasis::MIP, 4);

      if (cplex.solve()) {
          cout << "Solution status: " << cplex.getStatus() << endl;
          cout << " Optimal Value = " << cplex.getObjValue() << endl;
      }
   }
   catch (IloException& ex) {
      cerr << "Error: " << ex << endl;
   }
   catch (...) {
      cerr << "Error" << endl;
   }
   env.end();
   return 0;
}

/* 
Example input file:
[[1, 3, 4, 1, 2, 4, 2, 4],
 [3, 0, 0, 3, 0, 3, 4, 2],
 [0, 1, 2, 4, 1, 2, 0, 1],
 [4, 4, 1, 2, 3, 0, 3, 0],
 [2, 2, 3, 0, 4, 1, 1, 3],
 [6, 5, 6, 6, 5, 6, 5, 5],
 [7, 6, 7, 7, 7, 7, 7, 7],
 [5, 7, 5, 5, 6, 5, 6, 6]]
[[41, 32, 72, 65, 53, 35, 53, 2],
 [21, 44, 75, 58, 66, 7, 9, 16],
 [21, 70, 3, 85, 70, 39, 86, 96],
 [95, 78, 57, 65, 57, 58, 35, 9],
 [35, 34, 5, 34, 1, 12, 11, 86],
 [73, 98, 22, 56, 31, 40, 53, 71],
 [42, 95, 27, 17, 61, 92, 30, 66],
 [70, 49, 13, 71, 77, 5, 34, 99]]
[983, 1017, 1353, 1071, 1211, 1039, 1368, 1139]
[3, 9, 13, 19, 20, 20, 6, 9]
[10, 11, 4, 12, 15, 7, 6, 5]
*/

File: cplex/examples/src/java/Etsp.java
/* --------------------------------------------------------------------------
 * File: Etsp.java
 * Version 12.6.1  
 * --------------------------------------------------------------------------
 * Licensed Materials - Property of IBM
 * 5725-A06 5725-A29 5724-Y48 5724-Y49 5724-Y54 5724-Y55 5655-Y21
 * Copyright IBM Corporation 2001, 2014. All Rights Reserved.
 *
 * US Government Users Restricted Rights - Use, duplication or
 * disclosure restricted by GSA ADP Schedule Contract with
 * IBM Corp.
 * --------------------------------------------------------------------------
 *
 * Etsp.java - Solving an earliness-tardiness scheduling problem
 *             using CPLEX linearization capabilities.
 *
 * A command line argument indicating the input data file is required
 * to run this example.
 *
 *     java Etsp ../../../examples/data/etsp.dat
 */

import ilog.concert.*;
import ilog.cplex.*;

public class Etsp {
   static private int Horizon = 10000;

   static private class Data {
      int        nJobs;
      int        nResources;
      int[][]    activityOnResource;
      double[][] duration;
      double[]   dueDate;
      double[]   earlinessCost;
      double[]   tardinessCost;

      Data(String filename) throws IloException, java.io.IOException,
                                   InputDataReader.InputDataReaderException
      {
         InputDataReader reader = new InputDataReader(filename);

         activityOnResource = reader.readIntArrayArray();
         duration           = reader.readDoubleArrayArray();
         dueDate            = reader.readDoubleArray();
         earlinessCost      = reader.readDoubleArray();
         tardinessCost      = reader.readDoubleArray();

         nJobs      = dueDate.length;
         nResources = activityOnResource.length;
      }
   }

   public static void main (String args[]) {
      try {
         String filename;
         if ( args.length > 1)  filename = args[0];
         else                   filename = "../../../examples/data/etsp.dat";

         Data     data  = new Data(filename);
         IloCplex cplex = new IloCplex();

         // Create start variables
         IloNumVar[][] s = new IloNumVar[data.nJobs][];
         for (int j = 0; j < data.nJobs; j++)
            s[j] = cplex.numVarArray(data.nResources, 0.0, Horizon);

         // State precedence constraints
         for (int j = 0; j < data.nJobs; j++) {
            for (int i = 1; i < data.nResources; i++)
               cplex.addGe(s[j][i], cplex.sum(s[j][i-1], data.duration[j][i-1]));
         }

         // State disjunctive constraints for each resource
         for (int i = 0; i < data.nResources; i++) {
            int end = data.nJobs - 1;
            for (int j = 0; j < end; j++) {
               int a = data.activityOnResource[i][j];
               for (int k = j + 1; k < data.nJobs; k++) {
                  int b = data.activityOnResource[i][k];
                  cplex.add(cplex.or(
                     cplex.ge(s[j][a], cplex.sum(s[k][b], data.duration[k][b])),
                     cplex.ge(s[k][b], cplex.sum(s[j][a], data.duration[j][a]))
                  ));
               }
            }
         }

         // The cost is the sum of earliness or tardiness costs of each job 
         int last = data.nResources - 1;
         IloNumExpr costSum = cplex.numExpr();
         for (int j = 0; j < data.nJobs; j++) {
            double[] points = { data.dueDate[j] };
            double[] slopes = { -data.earlinessCost[j],
                                data.tardinessCost[j] };
            costSum = cplex.sum(costSum,
               cplex.piecewiseLinear(
                  cplex.sum(s[j][last], data.duration[j][last]),
                  points, slopes, data.dueDate[j], 0)
            );
         }
         cplex.addMinimize(costSum);

         cplex.setParam(IloCplex.Param.Emphasis.MIP, 4);

         if ( cplex.solve() ) {
             System.out.println("Solution status: " + cplex.getStatus());
             System.out.println(" Optimal Value = " + cplex.getObjValue());
         }
         cplex.end();
      }
      catch (IloException e) {
         System.err.println("Concert exception caught: " + e);
      }    
      catch (InputDataReader.InputDataReaderException ex) {
         System.out.println("Data Error: " + ex);
      }
      catch (java.io.IOException ex) {
         System.out.println("IO Error: " + ex);
      }
   }
}

